src/alert_manager.cpp

```diff
@@ -152,54 +175,147 @@ namespace libtorrent
 	void alert_manager::set_notify_function(boost::function<void()> const& fun)
 	{
 		mutex::scoped_lock lock(m_mutex);
-		m_notify = fun;
-		if (!m_alerts[m_generation].empty())
+
+		// in order to make boost:function<> atomic we use two separate
+		// fields in a ping-pong fashion and update the pointer to it
+		// atomically
+		if (fun)
+		{
+			const int index = (m_notify == &m_notify_storage[0]) ? 1 : 0;
+			m_notify_storage[index] = fun;
+			m_notify.store(&m_notify_storage[index]);
+		}
+		else
+		{
+			m_notify.store(NULL);
+		}
+
+		if (m_queue_size != 0)
 		{
 			// never call a callback with the lock held!
-			lock.unlock();
-			if (m_notify) m_notify();
+			if (m_notify)
+				(*m_notify.load(boost::memory_order_relaxed))();
```

Comment:
No, because the load in the if statement is not relaxed. It would actually be a race condition if we make it non-relaxed. It would be a race condition if the load was not relaxed because it can change after the first load.

There may still be a bug here because I'm not sure if the compiler can decide to still do the second load. So it should be loaded to a local variable first.

Reply:
Actually, I thought I was looking at the same call on `do_enqueue_alert()`. In this case there's no problem and even the null check should be a relaxed load. The reason is that this is the only location that updates `m_notify` and it is done under the mutex, so the `m_notify` store just a few lines above guarantees that we don't have a stale value. The only reason that the `m_notify` store needs to be synchronized is to guarantee that threads calling `emplace_alert()` will see the new pointer after `m_notify_storage` is set. Within this function only single threaded rules apply.

Reply:
it looks like this code has changed now, so I'm not referencing the code. But this is done after releasing the mutex. So all bets are off again, no? the underlying storage of the function object may have been written to by the time we get here.

Reply:
You mean if another thread calls the same function. Yes, I see it. It's not just on my code but it exists on RC_1_1 too. I think the correct solution is to invoke `fun()` instead of `m_notify()` in here.

On RC_1_1 you also have the same race-condition on `maybe_notify()`, so it's really part of the existing bug report. On this commit that part is fixed by the shared lock, as `set_dispatch_function` or `set_notify_function` is not called from the callbacks, that would result in deadlock, everything else is OK.